src/ray/core_worker/reference_count.cc

```diff
@@ -1120,16 +1139,41 @@ Status ReferenceCounter::SubscribeObjectLocations(
     // If the last location version is less than the current location version, we
     // already have location data that the subscriber hasn't seen yet, so we immediately
     // invoke the callback.
-    callback(it->second.locations, it->second.object_size, it->second.spilled_url,
-             it->second.spilled_node_id, it->second.location_version,
-             it->second.pinned_at_raylet_id);
-  } else {
-    // Otherwise, save the callback for later invocation.
-    it->second.location_subscription_callbacks.push_back(callback);
+    PublishObjectLocations(object_id, it->second.locations, it->second.object_size,
+                           it->second.spilled_url, it->second.spilled_node_id,
+                           it->second.location_version, it->second.pinned_at_raylet_id);
   }
   return Status::OK();
 }
 
+void ReferenceCounter::PublishObjectLocations(
+    const ObjectID &object_id, const absl::flat_hash_set<NodeID> &locations,
+    int64_t object_size, const std::string &spilled_url, const NodeID &spilled_node_id,
+    int64_t current_version, const absl::optional<NodeID> &optional_primary_node_id) {
+  auto primary_node_id = optional_primary_node_id.value_or(NodeID::Nil());
+  RAY_LOG(DEBUG) << "Publish a message for " << object_id
+                 << " with location update version " << current_version << ", "
+                 << locations.size() << " locations, spilled url: " << spilled_url
+                 << ", spilled node ID: " << spilled_node_id
+                 << ", and object size: " << object_size
+                 << ", and primary node ID: " << primary_node_id;
+  rpc::PubMessage pub_message;
+  pub_message.set_key_id(object_id.Binary());
+  pub_message.set_channel_type(rpc::ChannelType::WORKER_OBJECT_LOCATIONS_CHANNEL);
+  auto object_locations_msg = pub_message.mutable_worker_object_locations_message();
+
+  for (const auto &node_id : locations) {
+    object_locations_msg->add_node_ids(node_id.Binary());
+  }
+  object_locations_msg->set_object_size(object_size);
+  object_locations_msg->set_spilled_url(spilled_url);
+  object_locations_msg->set_spilled_node_id(spilled_node_id.Binary());
+  object_locations_msg->set_current_version(current_version);
+  object_locations_msg->set_primary_node_id(primary_node_id.Binary());
+  object_status_publisher_->Publish(rpc::ChannelType::WORKER_OBJECT_LOCATIONS_CHANNEL,
```

Comment:
Nit: Might not be worth the change, but `object_status_publisher_` should probably be renamed to something like `object_info_publisher_` or something since it's scope now includes publishing object locations.

Reply:
Good point. Changed all publisher & subscribers to be object_info_[subscriber | publisher]